asterius/test/ghc-testsuite.hs

```diff
@@ -224,34 +234,94 @@ consoleOutput tlref toutput smap =
       , Any nonempty
       )
 
+-- | Filter the test tree according to a predicate
+filterTestTree :: (TestName -> Bool) -> TestTree -> TestTree
+filterTestTree f tt =
+  TestGroup "***filtered***" $ [go tt]
+  where
+    -- go :: TestTree -> TestTree
+    go (SingleTest name t) =
+      if (f name)
+        then SingleTest (name <> "-ENABLED") t
+        else TestGroup (name <> "-DISABLED") []
+    go (TestGroup name tt) =
+      TestGroup (name <> "-FILTERED") (map go tt)
+    go _ = error $ "unknown test tree type"
+
+-- | Option that describes whether a test file with
+-- white & black lists has been provided. Modelled after
+-- TestPattern
+newtype PatternFilePath = PatternFilePath (Maybe FilePath)
+  deriving(Typeable, Show, Eq)
```

Comment:
`Typeable` deriving is on by default since a long time ago, no need to make it explicit here.